rllib/core/rl_trainer/rl_trainer.py

```diff
@@ -0,0 +1,155 @@
+        post_processed_gradients: Mapping[str, Any],
+    ) -> Mapping[str, Any]:
+        """Compile results from the update.
+
+        Args:
+            batch: The batch that was used for the update.
+            fwd_out: The output of the forward train pass.
+            postprocessed_loss: The loss after postprocessing.
+            post_processed_gradients: The gradients after postprocessing.
+
+        Returns:
+            A dictionary of results.
+        """
+        loss_numpy = convert_to_numpy(postprocessed_loss)
+        batch = convert_to_numpy(batch)
+        return {
+            "avg_reward": batch["rewards"].mean(),
+            **loss_numpy,
+        }
+
+    @abc.abstractmethod
+    def update(self, batch: SampleBatch) -> Mapping[str, Any]:
+        """Perform an update on this Trainer.
+
+        Args:
+            batch: A batch of data.
+
+        Returns:
+            A dictionary of results.
+        """
+
+    @abc.abstractmethod
+    def compute_gradients(
+        self, loss: Union[TensorType, Mapping[str, Any]]
+    ) -> Mapping[str, Any]:
+        """Perform an update on self._module.
+
+        For example compute and apply gradients to self._module if
+        necessary.
+
+        Args:
+            loss: variable(s) used for optimizing self._module.
+
+        Returns:
+            A dictionary of extra information and statistics.
+        """
+
+    def set_state(self, state: Mapping[str, Any]) -> None:
+        """Set the state of the trainer."""
+        self._rl_optimizer.set_state(state.get("optimizer_state", {}))
+        self._module.set_state(state.get("module_state", {}))
+
+    def get_state(self) -> Mapping[str, Any]:
+        """Get the state of the trainer."""
+        return {
+            "module_state": self._module.get_state(),
+            "optimizer_state": self._rl_optimizer.get_state(),
+        }
+
+    def add_module(
+        self,
+        module_id: ModuleID,
+        module_cls: Type[RLModule],
+        module_config,
+        optimizer_cls,
+        optimizer_config,
+    ) -> None:
+        """Add a module to the trainer."""
+        module = module_cls.from_model_config(**module_config)
+        self._module.add_module(module_id, module)
+        optimizer = optimizer_cls(module, optimizer_config)
+        self._rl_optimizer.add_optimizer(module_id, optimizer)
+
+    def remove_module(self, module_id: ModuleID) -> None:
+        """Remove a module from the trainer."""
+        self._module.remove_module(module_id)
+        self._rl_optimizer.remove_optimizer(module_id)
+
+    def init_trainer(self) -> None:
+        """Initialize the model."""
+        if self.distributed:
+            self._module, self._rl_optimizer = self.make_distributed()
+        else:
+            self._module = self.module_class.from_model_config(
+                **self.module_config
+            ).as_multi_agent()
+            self._rl_optimizer = self.optimizer_class(
```

Comment:
How would this support MARL? The signatures that you are using are for SARL. by calling `as_multi_agent()` on them you just return a dict with `DEFAULT_POLICY_ID`.

Reply:
yes I need to figure this out by updating the RLOptimizer class to have another function like from_config as I do in the MARLOptimizer class.